File: app/src/main/java/org/schabi/newpipe/database/history/dao/SearchHistoryDAO.java
package org.schabi.newpipe.database.history.dao;

import androidx.annotation.Nullable;
import androidx.room.Dao;
import androidx.room.Query;

import org.schabi.newpipe.database.history.model.SearchHistoryEntry;

import java.util.List;

import io.reactivex.rxjava3.core.Flowable;

import static org.schabi.newpipe.database.history.model.SearchHistoryEntry.CREATION_DATE;
import static org.schabi.newpipe.database.history.model.SearchHistoryEntry.ID;
import static org.schabi.newpipe.database.history.model.SearchHistoryEntry.SEARCH;
import static org.schabi.newpipe.database.history.model.SearchHistoryEntry.SERVICE_ID;
import static org.schabi.newpipe.database.history.model.SearchHistoryEntry.TABLE_NAME;
import static org.schabi.newpipe.database.history.model.SearchHistoryEntry.BOOKMARK;

@Dao
public interface SearchHistoryDAO extends HistoryDAO<SearchHistoryEntry> {
    String ORDER_BY_CREATION_DATE = " ORDER BY " + BOOKMARK + " DESC," + CREATION_DATE + " DESC";
    String ORDER_BY_MAX_CREATION_DATE = " ORDER BY " + BOOKMARK + " DESC, "
            + "MAX(" + CREATION_DATE + ") DESC";

    @Query("SELECT * FROM " + TABLE_NAME
            + " WHERE " + ID + " = (SELECT MAX(" + ID + ") FROM " + TABLE_NAME + ")")
    @Nullable
    SearchHistoryEntry getLatestEntry();

    @Query("DELETE FROM " + TABLE_NAME)
    @Override
    int deleteAll();

    @Query("DELETE FROM " + TABLE_NAME + " WHERE " + SEARCH + " = :query")
    int deleteAllWhereQuery(String query);

    @Query("SELECT * FROM " + TABLE_NAME + ORDER_BY_CREATION_DATE)
    @Override
    Flowable<List<SearchHistoryEntry>> getAll();

    @Query("SELECT * FROM " + TABLE_NAME + " GROUP BY " + SEARCH
            + ORDER_BY_MAX_CREATION_DATE + " LIMIT :limit")
    Flowable<List<SearchHistoryEntry>> getUniqueEntries(int limit);

    @Query("SELECT * FROM " + TABLE_NAME
            + " WHERE " + SERVICE_ID + " = :serviceId" + ORDER_BY_CREATION_DATE)
    @Override
    Flowable<List<SearchHistoryEntry>> listByService(int serviceId);

    @Query("SELECT * FROM " + TABLE_NAME + " WHERE " + SEARCH + " LIKE :query || '%'"
            + " GROUP BY " + SEARCH + ORDER_BY_MAX_CREATION_DATE + " LIMIT :limit")
    Flowable<List<SearchHistoryEntry>> getSimilarEntries(String query, int limit);
}

File: app/src/main/java/org/schabi/newpipe/database/history/dao/SearchHistoryDAO.kt
/*
 * SPDX-FileCopyrightText: 2017-2021 NewPipe contributors <https://newpipe.net>
 * SPDX-FileCopyrightText: 2025 NewPipe e.V. <https://newpipe-ev.de>
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

package org.schabi.newpipe.database.history.dao

import androidx.room.Dao
import androidx.room.Query
import kotlinx.coroutines.flow.Flow
import org.schabi.newpipe.database.BasicDAO
import org.schabi.newpipe.database.history.model.SearchHistoryEntry

@Dao
interface SearchHistoryDAO : BasicDAO<SearchHistoryEntry> {

    @Query("SELECT * FROM search_history WHERE id = (SELECT MAX(id) FROM search_history)")
    suspend fun getLatestEntry(): SearchHistoryEntry?

    @Query("DELETE FROM search_history")
    override suspend fun deleteAll(): Int

    @Query("DELETE FROM search_history WHERE search = :query")
    suspend fun deleteAllWhereQuery(query: String): Int

    @Query("SELECT * FROM search_history ORDER BY creation_date DESC")
    override fun getAll(): Flow<List<SearchHistoryEntry>>

    @Query("SELECT search FROM search_history GROUP BY search ORDER BY MAX(creation_date) DESC LIMIT :limit")
    fun getUniqueEntries(limit: Int): Flow<List<String>>

    @Query("SELECT * FROM search_history WHERE service_id = :serviceId ORDER BY creation_date DESC")
    override fun listByService(serviceId: Int): Flow<List<SearchHistoryEntry>>

    @Query(
        """
        SELECT search FROM search_history WHERE search LIKE :query ||
        '%' GROUP BY search ORDER BY MAX(creation_date) DESC LIMIT :limit
        """
    )
    fun getSimilarEntries(query: String, limit: Int): Flow<List<String>>
}
